### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by vincentberenz**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description

This initiates a refactor on how all default configuration files for all pam / pam_mujoco related packages are handled.

Currently, each package installs upon call to "colcon  build" files in /opt/mpi-is .

This has issues:
- duplication of cmake code in CMakeLists.txt (for the copy of config files)
- first creation of the config folders must be done with sudo rights
- config files are global: if a user changes the files for her/himself, then she/he changes them for everybody

This packages:
- places in one places all the defaults config files
- has an install script which copy the config files in the home directory of the user ...
- ... except if called with sudo, in which case it copies them globally
- has c++ API (with python wrappers) that returns the path to the configuration folder (in the home user if it exists, otherwise the global folder) 

Next, various packages (context, pam_models, pam_interface, tennicam_client, pam_mujoco) will need to be updated to use these configuration files.

## How I Tested

On my machine:
- Ran the install script with and without sudo
- Called 'colcon build'  
- Called the python API with all possible cases (no config folder, only the user config folder, only the global folder, with both folder)

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

**Comment by vincentberenz** [comment]
@luator @simonguist 

Because tons of config files have been added and clutter things, here the lists of files to be reviewed:

Files to review:

##CMakeLists.txt
https://github.com/intelligent-soft-robots/pam_configuration/pull/1/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a

##install
https://github.com/intelligent-soft-robots/pam_configuration/pull/1/files#diff-1e142e6277b12b7e1110478a24caee8f006a9349e86970c890203d6266209463

##pam_configuration.hpp
https://github.com/intelligent-soft-robots/pam_configuration/pull/1/files#diff-664b0676a3563566f144e2123454a89519010a2c7a7738f4456e99debe528c76

##pam_configuration.cpp
https://github.com/intelligent-soft-robots/pam_configuration/pull/1/files#diff-e519a8f4f74b0c52e9161b3b1bef8531a1dc89cd6bed4c4fd4a7fad8f19f57d6

##wrappers.cpp
https://github.com/intelligent-soft-robots/pam_configuration/pull/1/files#diff-c2fcf63a5a0feb0ff18cf9db2b3f222edfe07e5d7595a7f694731feedd48c803

File: src/pam_configuration.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    std::filesystem::path user_config(homedir);
+    user_config /= subdir;
+    return user_config;
+}
+
+bool user_path_exists()
+{
+    {
+        std::filesystem::path p = get_user_path();
+        if (std::filesystem::is_directory(p))
```
Comment (luator) [inline src/pam_configuration.cpp:34]: This could be simplified to `return std::filesystem::is_directory(p)`.

**Review by luator — COMMENTED** [review]
Is it actually necessary to install the files to a fixed location outside of the normal package install destination?  Especially if we now get a function to get the path, they could maybe just be installed as normal package resources in `.../workspace/install/pam_config/share/...`

**Review by simonguist — APPROVED** [review]

File: src/pam_configuration.cpp (hunk 1 above)
Comment (vincentberenz) [reply src/pam_configuration.cpp:34]: There are two reasons why "false" would be returned:
- get_user_path() raise an exception ($HOME environment variable is not defined)
- p does not exist or is not a directory
With the code above, if get_user_path raises an exception, false is also returned

**Comment by vincentberenz** [comment]
> Is it actually necessary to install the files to a fixed location outside of the normal package install destination? Especially if we now get a function to get the path, they could maybe just be installed as normal package resources in `.../workspace/install/pam_config/share/...`

Yes when using colcon, but I'd like this to support also binaries install

File: src/pam_configuration.cpp (hunk 1 above)
Comment (luator) [reply src/pam_configuration.cpp:34]: The exception isn't caught at the moment.  I guess there is a `try` missing in line 31?

**Comment by luator** [comment]
Is the installation location fixed when using the binaries or can it be changed by the user?  In the latter case I see that it would be difficult.

**Comment by vincentberenz** [comment]
> Is the installation location fixed when using the binaries or can it be changed by the user? In the latter case I see that it would be difficult.

It will be fixed. Plan is that the binaries installation script will call the install script of this folder to install the configuration file (i.e. ~/.mpi-is/pam if called without sudo, /opt/mpi-is/pam if called with sudo)

**Comment by luator** [comment]
I meant the installation location of the binaries.  But anyway I'm not sure if what I had in mind would actually work.  I was thinking that if you install to `${CMAKE_INSTALL_PREFIX}/share` you can create a preprocessor constant with 

    add_definitions(-DDATA_PATH="${CMAKE_INSTALL_PATH}/share/...")

and then use this in the `get_path()` function.
However, this would only work, if `${CMAKE_INSTALL_PATH}` during build time is the same as the location where the package is finally installed to, which might not be the case.

**Comment by vincentberenz** [comment]
My answer "when using colcon, but I'd like this to support also binaries install" was vague, and maybe we do not talk about the same thing.

I meant that I would like the config files to be at the same place either when you compile from source via a colcon workspace or when you installed directly the binaries.

The binaries tar ball and the related make file (as generated by codewings) will install the config files in ~/.mpi-is/pam.

This is why I would also like the config files to be in the same folder when using colcon (i.e. not in the workspace/install folder).

Actually I plan to have the installation process of the binaries using the install script of this package (as colcon will do)

**Comment by luator** [comment]
I see.  I don't like so much that a call of `colcon build` results in files being created outside of the workspace directory but if it otherwise gets too complicated with the binary installation, I'm okay with it.

There is still the open discussion above on `pam_configuration.cpp` (not sure if you saw my response there), apart from this it looks good to me :+1:

**Pull request merged**